Modernize code: use std::optional instead of bool/outarg

sipa opened the pull request:
Built on top of #106.

darosior commented on bitcoin/test/fuzz/miniscript.cpp:
```diff
@@ -77,13 +77,12 @@ struct TestData {
     typedef CPubKey Key;
 
-    bool ToString(const Key& key, std::string& ret) const
+    std::optional<std::string> ToString(const Key& key, std::string& ret) const
```
You need to remove the out argument, too.

darosior reviewed (commented):
To make a comprehensive list of the missing call sites from these changes:
- `thresh` parsing
- `multi` parsing
- `miniscript_string` usage of `ToString`
- the `KeyConverter` in the unit tests

From [my branch which has this included in Bitcoin Core](https://github.com/darosior/bitcoin/commits/miniscript_updates) on which i could actually compile.
Looks good to me otherwise, nice cleanup!

sipa commented:
Rebased, and addressed comments.

I've also made a few additional changes; turning inline "&& (num = ParseScriptNum(...)) &&" conditions into separate lines.

darosior commented:
ACK 033238fd9b8126bce29446eaf018af2bdbc8c906

Checked the diff with my [Bitcoin Core branch](https://github.com/darosior/bitcoin/commits/miniscript_updates) that i tested quite a lot. I didn't look for more potentially missed called sites too hard.
